### PR TITLE
Add test to prove redundant calls to identical target providers

### PR DESCRIPTION
**Pull request by grandbora**
This test proves that service discovery creates multiple instances for identical target providers. 

I am working on https://github.com/prometheus/prometheus/issues/2191, as a starter I wanted to write a test that proves the behavior that we want to modify.

I am planning to follow this up with changes in the discovery package that reuses the TargetProvider instances when they are identical.

cc @brian-brazil  @fabxc

### CODE REVIEW

File: discovery/discovery_test.go [hunk 1]
```diff
@@ -83,3 +84,48 @@ func (s *mockSyncer) Sync(tgs []*config.TargetGroup) {
 		s.sync(tgs)
 	}
 }
+
+type mockTargetProvider struct {
+	callCount *int
+}
+
+func (tp mockTargetProvider) Run(ctx context.Context, up chan<- []*config.TargetGroup) {
+	*tp.callCount = *tp.callCount + 1
```
Comment (brian-brazil) [inline discovery/discovery_test.go]: You've got a race here

Comment (grandbora) [reply discovery/discovery_test.go]: I used an atomic update operation here. https://github.com/prometheus/prometheus/pull/3404/commits/fb7a6a758d81776b14fb24168916e1357d603bfc

Comment (grandbora) [reply discovery/discovery_test.go]: ping @brian-brazil

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**